### ISSUE TITLE
Enhance Interval arithmetic to support String type. 

### ISSUE THREAD

**Issue by mingmwang** [apache/arrow-datafusion#7098]
### Is your feature request related to a problem or challenge?

Currently DataFusion does not support Interval arithmetic on string type, and the stats `analyze` process will be failed during CBO stats estimation process.

I think maybe we can normalize the string intervals to numeric intervals first and leverage the numeric intervals to estimate the `selectivity`.

Init input Bound:
["aaa", "xyz123"] => [0, 1]

Estimated Bound (example)
["ccc", "ddd"] => [0.2, 0.3]

estimated selectivity:
0.1

The Normalization of the string intervals can be based on ascil code ordering, for any non-ascil char,  we can just return early.

**Comment by mingmwang** [comment]
@berkaysynnada @ozankabak 
Please share your thoughts on this.

**Comment by mingmwang** [comment]
And I think we can restrict the max length of strings that we want to support and calculate a weighted number to represent the string first and normalize to a float, some sample code generated by chatgpt:

```rust
     let mut char_map: HashMap<char, i32> = HashMap::new();
        let mut value = 0;
        for c in '0'..='9' {
            char_map.insert(c, value);
            value += 1;
        }
        for c in 'a'..='z' {
            char_map.insert(c, value);
            value += 1;
        }
        for c in 'A'..='Z' {
            char_map.insert(c, value);
            value += 1;
        }
        let str_len = interval.chars().count();
        let mut sum = 0.0;
        for (i, c) in interval1.chars().enumerate() {
            let char_value = *char_map.get(&c).unwrap() as f64;
            let weight = (i + 1) as f64 / str_len as f64;
            sum += char_value * weight;
        }
        let normalized_value = sum / (char_map.len() as f64 * str_len as f64);

```

**Comment by berkaysynnada** [comment]
Sounds good. However, I'm not entirely sure about the correctness of the weight calculation. Shouldn't the weight of the leading characters be higher (assuming the order of the strings is as in the dictionary)? Once the strings are accurately mapped to floats, I believe there wouldn't be any problem with calculating selectivity as you've proposed.

**Comment by mingmwang** [comment]
This will output:

```rust
normalized_vector [0.0, 0.16394189376465151, 0.35710980460519687, 0.7175591189901976, 0.9999993427698617, 1.0, 1.0]
```

**Comment by mingmwang** [comment]
> 



> Sounds good. However, I'm not entirely sure about the correctness of the weight calculation. Shouldn't the weight of the leading characters be higher (assuming the order of the strings is as in the dictionary)? Once the strings are accurately mapped to floats, I believe there wouldn't be any problem with calculating selectivity as you've proposed.

Yes, you are right. The `weigh` is calculated based on the position of the char.

**Comment by alamb** [comment]
Started https://github.com/apache/arrow-datafusion/issues/7882 to collect improvement to interval arithmetic
